### DISCUSSION TITLE
Sooo.... what shall we do?! 😃

### DISCUSSION THREAD

**Discussion by roaldarbol** [sneakers-the-rat/paper-feeds#2] [Ideas]
Hi @sneakers-the-rat (and others who will surely join in) ;-)
What do you envision this app to be doing? I'll try to think front-end here, then we can always worry about how to code it afterwards. :P

In my mind, I imagine something like this will be facing the user - here the most basic example:
1. A search bar. Like the one on there now, but here we can click and choose the journal, and it will be added to a list...
2. The list - here we can see which journals have been selected. 
3. Download buttons (OPML and CSV)

However, if we also want to add in tags like Dan suggested, we may need to re-evaluate how that would change the UI. Additionally, I think it would be great to use different operators with the tags (e.g. eLife, but NOT about AI).

We could also have a tab for generating single RSS feeds, i.e. where they are not put into lists, but are exposed in raw to the user.
(TBH, I don't know very much about how all this database stuff works, so I'm curious to learn a bit about it. :) )

I imagine that using `streamlit` would be great for creating a simple, but slick front-end.

**Comment by sneakers-the-rat** [comment]
Hell ya im stoked youre interested in this.

So yes from interface perspective, good way to start. On the same page with your thoughts here

I think we can try and keep the web part of it very light - we're not trying to recreate a paper browser/reader (yet). So I agree: main goal of web interface is to send someone on their way with a bundle of RSS links or an OPML file.

At the risk of repeating exactly what u said: To avoid needing to make a "shopping cart" that persists across sessions, we could structure the chooser s.t. there are checkboxes in that journal selection screen, and checking a row causes that row to be sticky and persist across future queries. Then we have a big "export" button at the bottom that pops out an OPML, ezpz. That sounds like a good v1 milestone (incl all the backend stuff that yno makes the feeds actually work).

One interface question I have is we definitely want to make a clear visual distinction between "journals that exist" and "journals that the server is tracking and generating feeds for" - I figure someone checking a box and exporting a link is enough of a signal that they want that journal indexed, do we need both a button for "create feed" and a checkbox? I ask just bc checkbox is weird semantics for a single action, but it might make more sense once we actually mock that up.

I think individual RSS feed pages would basically just be the feed metadata and a preview of the feed, that would be pretty easy to do with the same pattern the existing journal list is done, just by creating a `/feeds/{ISSN}` route. The feed itself could then be `/feeds/{ISSN}/rss` or something. Talked to @lambdaloop about this briefly last night.

Tag feed interface is more up in the air for me - I think the simplest thing would be to start with just making feeds for individual tags. Ill make an issue in a sec outlining how we might do that. I think getting into making feeds for Boolean queries could get pretty hectic pretty fast, but mostly from a caching/URL generation POV, so not too bad. Eg ideally the feed URL is 1:1 with the query, but we wouldnt want to have both `/feeds/not:ai+journal:elife` and `/feeds/journal:elife+not:ai` and even then putting the query in the URL is pretty cumbersome. I guess we could use query params on top of the rss feed like `feeds/{ISSN}/rss?not=tag:ai` and then the problem becomes a caching and interface prob. There's some line to draw between the work of the feed and the reader: eg. We shouldnt try and make compound feeds for multiple journals bc aggregating feeds is the whole point of RSS readers, and we could include keywords as item metadata, but I think those are later problems. For now lets keep it simple with individual tags and see what we want from there.

I had wanted to use this project as an excuse to explore htmx, which I have heard a lot about and love the idea of, which lets us do nice things like basically not having to program an independent frontend and all the api call boilerplate that entails, and also avoid needing to do a ton of pre-serve serverside rendering. I literally love the simplicity of this:

https://github.com/sneakers-the-rat/journal-rss/blob/23636f3fd78df215b92d05644d19e3dbbdbcc1a1/src/templates/index.html#L11-L24

Which gives us a complete API call and indicator with just html attrs, and being able to return like this:

https://github.com/sneakers-the-rat/journal-rss/blob/23636f3fd78df215b92d05644d19e3dbbdbcc1a1/src/journal_rss/app.py#L67

Which is just hydrating a jinja template with the result:

https://github.com/sneakers-the-rat/journal-rss/blob/main/src/templates/partials/feed-list.html

The "not shipping any js but htmx" is a fun goal, but not an ironclad one, so if we did want to persist data in local storage to make a shopping cart thing then thats np (I actually bet HTMX has some way of doing that)

So im not sure how that interacts with streamlit, which seems to want to be the server, but im ofc open to it if you want to play with that :)

**Comment by roaldarbol** [comment]
Coolio! I'll try to structure my thoughts.
First of all, I completely agree about the "Shopping cart" - I don't think we should cache between sessions. People can always add their missing feeds and subsequently export their complete OPML from their aggregator.

# 1. Overall UI Ideas
I've sketched out an idea for a layout and roughly the functions needed - I've added an image further down.
I imagine 3 overall sections:
1. Browse journals
2. View and customise selections
3. Download

### Browse journals
This is basically what is implemented now in `dev`. The difference is that rather than showing the RSS feed on the button click, it adds that journal and its metadata to a list - let's for fun call it `selected_feeds` - the button could well be a `+`. (And the same journal can be selected multiple times, which will be good for when we implement filtering).

### View and customise
This section shows the `selected_feeds` as a table (maybe with less metadata?). Instead of the `+` button, there's a `-` button in this one, so you can remove the entry from `selected_feeds`. 

Customisation is for v2, and my vision here is not super easy to implement, but I think a very slick solution. I envision a `Filter` button on each row. The button creates a pop-up, here I'm drawing a lot of design inspiration from the [obsidian-projects](https://github.com/marcusolsson/obsidian-projects) implementation (see pictures further down). 

### Download
The download button. I don't think the user needs to know that we create a feed - I imagine the functionality of "The Big Fat Button" being:
1. Check whether feeds exist.
2. Create feeds that don't exist
3. Export OPML

### Sketch
![IMG_2985 Large](https://github.com/sneakers-the-rat/journal-rss/assets/25629697/9d266314-b65d-4701-88f5-03e37f17c03f)

### Filtering in `obsidian-projects`
Filter button:
<img width="112" alt="Screenshot 2024-01-03 at 11 40 20" src="https://github.com/sneakers-the-rat/journal-rss/assets/25629697/0841a732-1803-417b-801f-e489b2dee934"><br>
Pop-up:
<img width="586" alt="Screenshot 2024-01-03 at 11 39 33" src="https://github.com/sneakers-the-rat/journal-rss/assets/25629697/aa834131-5262-41d6-83fb-9c42657fed67"> <br>
Choose variable:
<img width="594" alt="Screenshot 2024-01-03 at 11 39 54" src="https://github.com/sneakers-the-rat/journal-rss/assets/25629697/729bcbb1-074d-48bc-a854-68c7f26f3637"><br>
Choose logical operator:
<img width="584" alt="Screenshot 2024-01-03 at 11 39 43" src="https://github.com/sneakers-the-rat/journal-rss/assets/25629697/f5b86ec2-95c6-43dd-97bb-b027f0354dce"><br>

# 2. Styling decisions
I do like the idea of using htmx, however, without further JS it means we'll have to code the CSS up - which might be fine since the number of elements should remain quite low. It'll especially be the tables that are of importance, since they - at least for V2 - should have some extra functionality; I reckon we're better off trying to find a lightweight JS library that can do some of those things out of the box. I also think generally going with a Material Design would be nice, but am open to suggestions. There are some light CSS libraries for (e.g. https://getmdl.io/). The reason I'd be leaning in this direction is mostly my busyness and laziness.

**Reply by roaldarbol** [reply]
Okay, after having fiddled with the CSS, I think we can make that work - BUT I do think we'll need extra JS for the tables.
